syntax="proto3";

import "proto/users.proto";
import "proto/groups.proto";
import "proto/events.proto";

package distributed_agenda;
option go_package = "./server/proto";

enum Action{
  CREATE = 0;
  UPDATE = 1;
  DELETE = 2;
  JOINED = 3;
  LEFT = 4;
  CONFIRM = 5;
  REJECT = 6;
}

message HistoryEntry{
  Action action = 1;
  User user = 2;
  Group group = 3;
  Event event = 4;
}

message History{
  repeated HistoryEntry entries = 1;
}

message AddHistoryEntryRequest{
  HistoryEntry entry = 1;
  repeated string users = 2;
}

message AddHistoryEntryResponse{
}

message GetFullHistoryRequest{
  string username = 1;
}

message GetFullHistoryResponse{
  HistoryEntry entry = 1;
}

message GetHistoryFromOffsetRequest{
  string username = 1;
  int32 offset = 2;
}

message GetHistoryFromOffsetResponse{
  HistoryEntry entry = 1;
}

service HistoryService{
  rpc AddHistoryEntry(AddHistoryEntryRequest) returns (AddHistoryEntryResponse){}

  rpc GetFullHistory(GetFullHistoryRequest) returns (stream GetFullHistoryResponse){}

  rpc GetHistoryFromOffset(GetHistoryFromOffsetRequest) returns (stream GetHistoryFromOffsetResponse){}
}